### ISSUE TITLE
code unreachable rule

### ISSUE THREAD

**Issue by michelkluger** [microsoft/pyright#10284]
If you have a question about a behavior that you’re seeing in pyright, consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Is your feature request related to a problem? Please describe.**
pylance singals when code is unreachable would be nice to get a CI error via pyright

**Describe the solution you’d like**
A rule to report unreachable code

**Comment by erictraut** [comment]
This is a duplicate of [this issue](https://github.com/microsoft/pyright/issues/2080). It has received no upvotes in four years.

**Comment by michelkluger** [comment]
> This is a duplicate of [this issue](https://github.com/microsoft/pyright/issues/2080). It has received no upvotes in four years.

maybe would be nicer to implement it, I am sure many people and codebases will benefit from it, dont see what would be the negative of having such a rule

**Comment by michelkluger** [comment]
how many upvotes we need it to get it implemented? @erictraut

**Comment by erictraut** [comment]
As I mention in the other thread, there are many legitimate situations where unreachable code exists in a code base. That means there will be quite a few false positives for such a rule. It also means that such a rule could not be enabled by default even in strict mode. I generally don't like adding new rules that cannot be enabled by default even in strict mode because they are hard for users to discover, and therefore they rarely get used.

For those reasons, I'm not convinced this rule is a good idea. I'd need to see a lot of upvotes to be convinced otherwise.

**Comment by michelkluger** [comment]
> I'd need to see a lot of upvotes to be convinced otherwise.

how many?

**Comment by erictraut** [comment]
> how many?

That's sufficient :)

I've added a new diagnostic check called `reportUnreachable`. It is off by default even in strict mode. It reports a diagnostic for code that is determined to be structurally unreachable (e.g. after a `return` statement) _or_ unreachable via type analysis (a type guard that narrows the type of an expression to `Never`).

This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.402.

**Comment by Andrej730** [comment]
Isn't it a bit confusing description for the new rule?

> reportUnreachable [boolean or string, optional]: Generate or suppress diagnostics for code that is determined to be structurally unreachable or unreachable by type analysis. The default value for this setting is "none".

It seems to imply that it will report all cases of unreachable code, though it's only seems to be focused on structural unreachability.

```python
# pyright: strict
# pyright: reportUnreachable=true

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    print('hi')
else:
    # Not reported.
    print('test')

def test():
    raise Exception()
    # reported: error: Code is structurally unreachable
    print('test')

def test2():
    assert False
    # not reported
    print('test')

def test3():
    if True:
        print()
    else:
        # not reported.
        print('hi')
```

Update. It seems to be not just structural, though it still seems description doesn't specify when unreachable code wouldn't be reported.

```python
# pyright: strict
# pyright: reportUnreachable=true

from typing import Literal

def test4(x: Literal[True]):
    if x:
        print()
    else:
        # reported
        print('hi')

def test5(x: int):
    if isinstance(x, int):
        print()
    else:
        # reported
        print('hi')

def test6(x: int):
    assert not isinstance(x, int)
    # reported
    print('hi')
```

Also noticed a caveat trying out this rule - it conflicts with `assert_never`, so users won't be able to use `assert_never` and this rule.

```python

# pyright: strict
# pyright: reportUnreachable=true

from typing import assert_never, Literal

def test(x: Literal["A", "B"]):
    if x == "A":
        pass
    elif x == "B":
        pass
    else:
        # error: Type analysis indicates code is unreachable (reportUnreachable)
        assert_never(x)
```
